File: notebook-explorer/r/tests/testthat/test-client.R
BASE_URL <- "https://evagene.example"
PEDIGREE_ID <- "11111111-1111-1111-1111-111111111111"
SCRATCH_ID <- "99999999-9999-9999-9999-999999999999"
INDIVIDUAL_ID <- "22222222-2222-2222-2222-222222222222"
RELATIVE_ID <- "33333333-3333-3333-3333-333333333333"
DISEASE_ID <- "44444444-4444-4444-4444-444444444444"

scripted_gateway <- function(responses) {
  state <- new.env(parent = emptyenv())
  state$responses <- responses
  state$calls <- list()
  gateway <- function(request) {
    state$calls[[length(state$calls) + 1L]] <- request
    if (length(state$responses) == 0L) {
      stop(sprintf("Unexpected extra call: %s %s", request$method, request$url))
    }
    head <- state$responses[[1L]]
    state$responses <- state$responses[-1L]
    head
  }
  list(gateway = gateway, state = state)
}

response <- function(status, body = NULL, text = NULL) {
  if (!is.null(body)) {
    text <- jsonlite::toJSON(body, auto_unbox = TRUE, null = "null")
  }
  if (is.null(text)) text <- ""
  list(status = as.integer(status), body_string = text)
}

build_client <- function(gateway) {
  evagene_client(gateway, BASE_URL, "evg_test")
}

# ----------------------------------------------------------- get_pedigrees

test_that("get_pedigrees returns the list from the API", {
  rec <- scripted_gateway(list(response(200, list(list(id = PEDIGREE_ID)))))
  client <- build_client(rec$gateway)

  result <- client$get_pedigrees()

  expect_equal(length(result), 1L)
  expect_equal(result[[1]]$id, PEDIGREE_ID)
  expect_equal(rec$state$calls[[1]]$method, "GET")
  expect_equal(rec$state$calls[[1]]$url, paste0(BASE_URL, "/api/pedigrees"))
  expect_equal(unname(rec$state$calls[[1]]$headers[["X-API-Key"]]), "evg_test")
})

# ----------------------------------------------------------- run_risk

test_that("run_risk posts model and extra body fields", {
  rec <- scripted_gateway(list(response(200, list(cancer_risk = list()))))
  client <- build_client(rec$gateway)

  client$run_risk(
    PEDIGREE_ID, "TYRER_CUZICK",
    extra = list(age_at_menarche = 12L, parity = 0L)
  )

  call <- rec$state$calls[[1]]
  expect_equal(call$method, "POST")
  expect_equal(call$url, sprintf("%s/api/pedigrees/%s/risk/calculate", BASE_URL, PEDIGREE_ID))
  body <- jsonlite::fromJSON(call$body_json, simplifyVector = FALSE)
  expect_equal(body$model, "TYRER_CUZICK")
  expect_equal(body$age_at_menarche, 12)
  expect_equal(body$parity, 0)
})

test_that("run_risk raises api_error on non-2xx", {
  rec <- scripted_gateway(list(response(500, list())))
  client <- build_client(rec$gateway)

  expect_error(client$run_risk(PEDIGREE_ID, "NICE"), class = "api_error")
})

# ----------------------------------------------------------- clone sequence

test_that("clone_pedigree_for_exploration sequences export / create / import", {
  rec <- scripted_gateway(list(
    response(200, text = "0 HEAD\n0 TRLR\n"),
    response(201, list(id = SCRATCH_ID)),
    response(204, list())
  ))
  client <- build_client(rec$gateway)

  scratch_id <- client$clone_pedigree_for_exploration(PEDIGREE_ID, "2026-04-20 12:00")

  expect_equal(scratch_id, SCRATCH_ID)
  expect_equal(length(rec$state$calls), 3L)
  expect_equal(rec$state$calls[[1]]$method, "GET")
  expect_equal(
    rec$state$calls[[1]]$url,
    sprintf("%s/api/pedigrees/%s/export.ged", BASE_URL, PEDIGREE_ID)
  )
  expect_equal(rec$state$calls[[2]]$method, "POST")
  expect_equal(rec$state$calls[[2]]$url, paste0(BASE_URL, "/api/pedigrees"))
  create_body <- jsonlite::fromJSON(rec$state$calls[[2]]$body_json, simplifyVector = FALSE)
  expect_true(startsWith(create_body$display_name, "[scratch] notebook-explorer"))

  expect_equal(rec$state$calls[[3]]$method, "POST")
  expect_equal(
    rec$state$calls[[3]]$url,
    sprintf("%s/api/pedigrees/%s/import/gedcom", BASE_URL, SCRATCH_ID)
  )
  import_body <- jsonlite::fromJSON(rec$state$calls[[3]]$body_json, simplifyVector = FALSE)
  expect_equal(import_body$content, "0 HEAD\n0 TRLR\n")
})

test_that("clone_pedigree_for_exploration rejects create without id", {
  rec <- scripted_gateway(list(
    response(200, text = "0 HEAD\n0 TRLR\n"),
    response(201, list(name = "no id"))
  ))
  client <- build_client(rec$gateway)

  expect_error(
    client$clone_pedigree_for_exploration(PEDIGREE_ID, "2026-04-20 12:00"),
    class = "api_error"
  )
})

# ----------------------------------------------------------- delete + mutate

test_that("delete_pedigree issues DELETE", {
  rec <- scripted_gateway(list(response(204, list())))
  build_client(rec$gateway)$delete_pedigree(SCRATCH_ID)

  expect_equal(rec$state$calls[[1]]$method, "DELETE")
  expect_equal(
    rec$state$calls[[1]]$url,
    sprintf("%s/api/pedigrees/%s", BASE_URL, SCRATCH_ID)
  )
})

test_that("add_relative posts to register endpoint", {
  rec <- scripted_gateway(list(response(200, list(individual = list(id = RELATIVE_ID)))))
  client <- build_client(rec$gateway)

  result <- client$add_relative(
    SCRATCH_ID,
    relative_of = INDIVIDUAL_ID,
    relative_type = "sister",
    display_name = "Scratch sister",
    biological_sex = "female"
  )

  expect_equal(result$individual$id, RELATIVE_ID)
  call <- rec$state$calls[[1]]
  expect_equal(
    call$url,
    sprintf("%s/api/pedigrees/%s/register/add-relative", BASE_URL, SCRATCH_ID)
  )
  body <- jsonlite::fromJSON(call$body_json, simplifyVector = FALSE)
  expect_equal(body$relative_type, "sister")
  expect_equal(body$biological_sex, "female")
})

test_that("add_disease_to_individual posts affection and age", {
  rec <- scripted_gateway(list(response(201, list())))
  build_client(rec$gateway)$add_disease_to_individual(
    RELATIVE_ID, disease_id = DISEASE_ID, age_at_diagnosis = 42L
  )

  call <- rec$state$calls[[1]]
  expect_equal(call$url, sprintf("%s/api/individuals/%s/diseases", BASE_URL, RELATIVE_ID))
  body <- jsonlite::fromJSON(call$body_json, simplifyVector = FALSE)
  expect_equal(body$disease_id, DISEASE_ID)
  expect_equal(body$affection_status, "affected")
  expect_equal(body$age_at_diagnosis, 42)
})

test_that("patch_individual PATCHes the individual with arbitrary fields", {
  rec <- scripted_gateway(list(response(200, list())))
  build_client(rec$gateway)$patch_individual(
    INDIVIDUAL_ID,
    fields = list(age_at_menarche = 12L, parity = 2L, breast_density_birads = 3L)
  )

  call <- rec$state$calls[[1]]
  expect_equal(call$method, "PATCH")
  expect_equal(call$url, sprintf("%s/api/individuals/%s", BASE_URL, INDIVIDUAL_ID))
  body <- jsonlite::fromJSON(call$body_json, simplifyVector = FALSE)
  expect_equal(body$age_at_menarche, 12)
  expect_equal(body$parity, 2)
  expect_equal(body$breast_density_birads, 3)
})

test_that("get_register returns server object", {
  rec <- scripted_gateway(list(response(200, list(proband_id = INDIVIDUAL_ID, rows = list()))))
  result <- build_client(rec$gateway)$get_register(SCRATCH_ID)

  expect_equal(result$proband_id, INDIVIDUAL_ID)
  call <- rec$state$calls[[1]]
  expect_equal(call$method, "GET")
  expect_equal(
    call$url,
    sprintf("%s/api/pedigrees/%s/register", BASE_URL, SCRATCH_ID)
  )
})

# ----------------------------------------------------------- evagene_url

test_that("evagene_url is credential free", {
  rec <- scripted_gateway(list())
  url <- build_client(rec$gateway)$evagene_url(PEDIGREE_ID)

  expect_equal(url, sprintf("%s/pedigrees/%s", BASE_URL, PEDIGREE_ID))
  expect_false(grepl("evg_", url))
})

File: notebook-explorer/python/tests/test_client.py
from dataclasses import dataclass, field
from typing import Any

import pytest

from notebook_explorer.client import ApiError, EvageneClient

_BASE_URL = "https://evagene.example"
_PEDIGREE_ID = "11111111-1111-1111-1111-111111111111"
_SCRATCH_ID = "99999999-9999-9999-9999-999999999999"
_INDIVIDUAL_ID = "22222222-2222-2222-2222-222222222222"
_RELATIVE_ID = "33333333-3333-3333-3333-333333333333"
_DISEASE_ID = "44444444-4444-4444-4444-444444444444"


@dataclass
class _Call:
    method: str
    url: str
    headers: dict[str, str]
    json_body: dict[str, Any] | None


class _StubResponse:
    def __init__(
        self,
        status_code: int,
        payload: Any = None,
        text: str = "",
    ) -> None:
        self.status_code = status_code
        self._payload = payload
        self._text = text

    def json(self) -> Any:
        return self._payload

    @property
    def text(self) -> str:
        return self._text


@dataclass
class _ScriptedGateway:
    """Replies with a queue of responses; records every request for inspection."""

    responses: list[_StubResponse]
    calls: list[_Call] = field(default_factory=list)

    def request(
        self,
        method: str,
        url: str,
        *,
        headers: dict[str, str],
        json_body: dict[str, Any] | None = None,
    ) -> _StubResponse:
        self.calls.append(_Call(method=method, url=url, headers=headers, json_body=json_body))
        if not self.responses:
            raise AssertionError(f"Unexpected extra call: {method} {url}")
        return self.responses.pop(0)


def _client(gateway: _ScriptedGateway) -> EvageneClient:
    return EvageneClient(
        base_url=_BASE_URL,
        api_key="evg_test",
        http=gateway,
        rate_limit_sleeper=lambda _seconds: None,
        rate_limit_wait_seconds=0.0,
    )


# ----------------------------------------------------------- get_pedigrees


def test_get_pedigrees_returns_the_list_from_the_api() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(200, [{"id": _PEDIGREE_ID}])])

    pedigrees = _client(gateway).get_pedigrees()

    assert pedigrees == [{"id": _PEDIGREE_ID}]
    call = gateway.calls[0]
    assert call.method == "GET"
    assert call.url == f"{_BASE_URL}/api/pedigrees"
    assert call.headers["X-API-Key"] == "evg_test"


def test_get_pedigrees_rejects_non_list_payload() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(200, {"not": "a list"})])

    with pytest.raises(ApiError):
        _client(gateway).get_pedigrees()


# ----------------------------------------------------------- run_risk


def test_run_risk_posts_model_and_extra_body() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(200, {"cancer_risk": {}})])

    _client(gateway).run_risk(_PEDIGREE_ID, "TYRER_CUZICK", age_at_menarche=12, parity=0)

    call = gateway.calls[0]
    assert call.method == "POST"
    assert call.url == f"{_BASE_URL}/api/pedigrees/{_PEDIGREE_ID}/risk/calculate"
    assert call.json_body == {"model": "TYRER_CUZICK", "age_at_menarche": 12, "parity": 0}


def test_run_risk_rejects_non_object_payload() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(200, ["nope"])])

    with pytest.raises(ApiError):
        _client(gateway).run_risk(_PEDIGREE_ID, "NICE")


def test_run_risk_raises_on_http_error() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(500, {})])

    with pytest.raises(ApiError):
        _client(gateway).run_risk(_PEDIGREE_ID, "NICE")


# ----------------------------------------------------------- clone sequence


def test_clone_pedigree_sequences_export_create_import() -> None:
    gateway = _ScriptedGateway(
        responses=[
            _StubResponse(200, text="0 HEAD\n0 TRLR\n"),
            _StubResponse(201, {"id": _SCRATCH_ID}),
            _StubResponse(204, {}),
        ],
    )

    scratch_id = _client(gateway).clone_pedigree_for_exploration(
        _PEDIGREE_ID, scratch_suffix="2026-04-20 12:00"
    )

    assert scratch_id == _SCRATCH_ID
    assert [(call.method, call.url) for call in gateway.calls] == [
        ("GET", f"{_BASE_URL}/api/pedigrees/{_PEDIGREE_ID}/export.ged"),
        ("POST", f"{_BASE_URL}/api/pedigrees"),
        ("POST", f"{_BASE_URL}/api/pedigrees/{_SCRATCH_ID}/import/gedcom"),
    ]
    create_body = gateway.calls[1].json_body
    assert create_body is not None
    assert create_body["display_name"].startswith("[scratch] notebook-explorer")
    import_body = gateway.calls[2].json_body
    assert import_body == {"content": "0 HEAD\n0 TRLR\n"}


def test_clone_pedigree_rejects_create_response_without_id() -> None:
    gateway = _ScriptedGateway(
        responses=[
            _StubResponse(200, text="0 HEAD\n0 TRLR\n"),
            _StubResponse(201, {"name": "no id here"}),
        ],
    )

    with pytest.raises(ApiError):
        _client(gateway).clone_pedigree_for_exploration(
            _PEDIGREE_ID, scratch_suffix="2026-04-20 12:00"
        )


# ----------------------------------------------------------- delete + mutate


def test_delete_pedigree_issues_delete_request() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(204, {})])

    _client(gateway).delete_pedigree(_SCRATCH_ID)

    call = gateway.calls[0]
    assert call.method == "DELETE"
    assert call.url == f"{_BASE_URL}/api/pedigrees/{_SCRATCH_ID}"


def test_add_relative_posts_to_register_endpoint() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(200, {"individual": {"id": _RELATIVE_ID}})])

    result = _client(gateway).add_relative(
        _SCRATCH_ID,
        relative_of=_INDIVIDUAL_ID,
        relative_type="sister",
        display_name="Scratch sister",
        biological_sex="female",
    )

    assert result == {"individual": {"id": _RELATIVE_ID}}
    call = gateway.calls[0]
    assert call.url == f"{_BASE_URL}/api/pedigrees/{_SCRATCH_ID}/register/add-relative"
    assert call.json_body == {
        "relative_of": _INDIVIDUAL_ID,
        "relative_type": "sister",
        "display_name": "Scratch sister",
        "biological_sex": "female",
    }


def test_add_disease_to_individual_posts_affection_and_age() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(201, {})])

    _client(gateway).add_disease_to_individual(
        _RELATIVE_ID,
        disease_id=_DISEASE_ID,
        age_at_diagnosis=42,
    )

    call = gateway.calls[0]
    assert call.url == f"{_BASE_URL}/api/individuals/{_RELATIVE_ID}/diseases"
    assert call.json_body == {
        "disease_id": _DISEASE_ID,
        "affection_status": "affected",
        "age_at_diagnosis": 42,
    }


# ----------------------------------------------------------- evagene_url


def test_evagene_url_is_credential_free() -> None:
    gateway = _ScriptedGateway(responses=[])

    url = _client(gateway).evagene_url(_PEDIGREE_ID)

    assert url == f"{_BASE_URL}/pedigrees/{_PEDIGREE_ID}"
    assert "evg_" not in url


def test_add_disease_to_pedigree_posts_to_working_set() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(201, {})])

    _client(gateway).add_disease_to_pedigree(_SCRATCH_ID, _DISEASE_ID)

    call = gateway.calls[0]
    assert call.method == "POST"
    assert call.url == f"{_BASE_URL}/api/pedigrees/{_SCRATCH_ID}/diseases/{_DISEASE_ID}"


def test_patch_individual_sends_fields_as_body() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(200, {})])

    _client(gateway).patch_individual(
        _INDIVIDUAL_ID, age_at_menarche=12, parity=2, breast_density_birads=3
    )

    call = gateway.calls[0]
    assert call.method == "PATCH"
    assert call.url == f"{_BASE_URL}/api/individuals/{_INDIVIDUAL_ID}"
    assert call.json_body == {
        "age_at_menarche": 12,
        "parity": 2,
        "breast_density_birads": 3,
    }


def test_retries_on_rate_limit_until_success() -> None:
    gateway = _ScriptedGateway(
        responses=[
            _StubResponse(429, {}),
            _StubResponse(429, {}),
            _StubResponse(200, [{"id": _PEDIGREE_ID}]),
        ],
    )

    pedigrees = _client(gateway).get_pedigrees()

    assert pedigrees == [{"id": _PEDIGREE_ID}]
    assert len(gateway.calls) == 3


def test_retries_on_rate_limit_honour_max_retries() -> None:
    gateway = _ScriptedGateway(responses=[_StubResponse(429, {})] * 20)

    client = EvageneClient(
        base_url=_BASE_URL,
        api_key="evg_test",
        http=gateway,
        rate_limit_sleeper=lambda _seconds: None,
        rate_limit_wait_seconds=0.0,
        rate_limit_max_retries=3,
    )

    with pytest.raises(ApiError, match="rate-limited"):
        client.get_pedigrees()
    assert len(gateway.calls) == 4  # one initial attempt + 3 retries


def test_get_register_returns_server_object() -> None:
    gateway = _ScriptedGateway(
        responses=[_StubResponse(200, {"proband_id": _INDIVIDUAL_ID, "rows": []})]
    )

    result = _client(gateway).get_register(_SCRATCH_ID)

    assert result["proband_id"] == _INDIVIDUAL_ID
    call = gateway.calls[0]
    assert call.method == "GET"
    assert call.url == f"{_BASE_URL}/api/pedigrees/{_SCRATCH_ID}/register"
